ios/ReactNativeShareExtension.m

```diff
@@ -39,69 +39,89 @@ - (void)viewDidLoad {
-   NSDictionary *inventory = @{
-     @"type": contentType,
-     @"value": [url absoluteString]
-   };
-
-   resolve(inventory);
- }];
+  [self extractDataFromContext: extensionContext withCallback:^(NSDictionary *inventory ,NSException* err) {
+    resolve(inventory);
+  }];
 }
 
-- (void)extractDataFromContext:(NSExtensionContext *)context withCallback:(void(^)(NSURL *url, NSString* contentType ,NSException *exception))callback {
+- (void)extractDataFromContext:(NSExtensionContext *)context withCallback:(void(^)(NSDictionary *dict, NSException *exception))callback {
  @try {
    NSExtensionItem *item = [context.inputItems firstObject];
    NSArray *attachments = item.attachments;
    __block NSItemProvider *urlProvider = nil;
    __block NSItemProvider *imageProvider = nil;
+   
    [attachments enumerateObjectsUsingBlock:^(NSItemProvider *provider, NSUInteger idx, BOOL *stop) {
      if([provider hasItemConformingToTypeIdentifier:ITEM_IDENTIFIER]) {
        urlProvider = provider;
        *stop = YES;
-     }else if ([provider hasItemConformingToTypeIdentifier:IMAGE_IDENTIFIER]){
-         imageProvider = provider;
-         *stop = YES;
-
+     }
+     else if ([provider hasItemConformingToTypeIdentifier:IMAGE_IDENTIFIER]) {
+       imageProvider = provider;
+       *stop = YES;
      }
    }];
+     
+   if (!callback) {
+     return;
+   }
 
    if(urlProvider) {
      [urlProvider loadItemForTypeIdentifier:ITEM_IDENTIFIER options:nil completionHandler:^(id<NSSecureCoding> item, NSError *error) {
        NSURL *url = (NSURL *)item;
-
-       if(callback) {
-         callback(url,@"text/plain" ,nil);
-       }
+       callback([self createTypeValuePair:url contentType:@"text/plain"], nil);
      }];
-   }else if (imageProvider){
+   }
+   else if (imageProvider){
        [imageProvider loadItemForTypeIdentifier:IMAGE_IDENTIFIER options:nil completionHandler:^(id<NSSecureCoding> item, NSError *error) {
-           NSURL *url = (NSURL *)item;
-
-           if(callback) {
-               callback(url,[[[url absoluteString] pathExtension] lowercaseString] ,nil);
-           }
+         callback([self createImageDataObject:item], nil);
        }];
-
-
    }
    else {
-     if(callback) {
-       callback(nil, nil,[NSException exceptionWithName:@"Error" reason:@"couldn't find provider" userInfo:nil]);
-     }
+     callback(nil, [NSException exceptionWithName:@"Error" reason:@"Couldn't find provider" userInfo:nil]);
    }
  }
  @catch (NSException *exception) {
-   if(callback) {
-     callback(nil,nil ,exception);
-   }
+   callback(nil, exception);
  }
 }
 
+
+- (NSDictionary*) createTypeValuePair:(NSURL*) url
+                          contentType:(NSString*) contentType
+{
+  return @{
+           @"type": contentType,
+           @"value": [url absoluteString]
+           };
+}
+
+- (NSDictionary*) createImageDataObject:(NSURL*) url
+{
+  NSData *imageData = [NSData dataWithContentsOfURL:url];
```

Comment:
If your image file is large then reading it in to memory and decoding it into UI image may cause the extension to crash, you can get the metadata using ImageIO without having to load the image in to memory, this blog has an example: http://www.brians-brain.org/pholio/pholio-coders-guide/tips-for-effective-resizing/